mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputFormat.java

```diff
@@ -19,21 +19,408 @@
+
+    @Override
+    public void setupJob(JobContext jobContext) {
+      // do nothing.
+    }
+
+    @Override
+    public void setupTask(TaskAttemptContext taskAttemptContext) {
+      // do nothing.
+    }
+
+    @Override
+    public boolean needsTaskCommit(TaskAttemptContext taskAttemptContext) {
+      return true;
+    }
+
+    @Override
+    public void commitTask(TaskAttemptContext taskAttemptContext) throws IOException {
+      ClosedFileData closedFileData = fileData.remove(taskAttemptContext.getTaskAttemptID().toString());
+      createCommittedFileFor(new HadoopFileIO(taskAttemptContext.getJobConf()), closedFileData);
+    }
+
+    @Override
+    public void abortTask(TaskAttemptContext taskAttemptContext) {
+      fileData.remove(taskAttemptContext.getTaskAttemptID().toString());
+    }
+
+    @Override
+    public void commitJob(JobContext jobContext) throws IOException {
+      Configuration conf = jobContext.getJobConf();
+      Path queryResultPath = new Path(generateQueryLocation(conf));
+      Table table = Catalogs.loadTable(conf);
+
+      ExecutorService executor = null;
+      try {
+        // Creating executor service for parallel handling of file reads and deletes
+        executor = Executors.newFixedThreadPool(
+            conf.getInt(InputFormatConfig.WRITE_THREAD_POOL_SIZE, InputFormatConfig.WRITE_THREAD_POOL_SIZE_DEFAULT),
+            new ThreadFactoryBuilder()
+                .setDaemon(false)
+                .setPriority(Thread.NORM_PRIORITY)
+                .setNameFormat("iceberg-commit-pool-%d")
+                .build());
+
+        Set<String> taskTmpFiles = new HashSet<>();
+        Set<Future<DataFile>> dataFiles = new HashSet<>();
+        FileSystem fs = Util.getFs(queryResultPath, conf);
+
+        // Listing the task result directory and reading .committed files
+        RemoteIterator<LocatedFileStatus> taskFileStatuses = fs.listFiles(queryResultPath, false);
+        while (taskFileStatuses.hasNext()) {
+          LocatedFileStatus taskFile = taskFileStatuses.next();
+          String taskFileName = queryResultPath + "/" + taskFile.getPath().getName();
+          taskTmpFiles.add(taskFileName);
+          if (taskFileName.endsWith(COMMITTED_EXTENSION)) {
+            dataFiles.add(executor.submit(() -> {
+              LOG.debug("Reading committed file {}", taskFileName);
+              ClosedFileData cfd = readCommittedFile(table.io(), taskFileName);
+              DataFiles.Builder builder = DataFiles.builder(PartitionSpec.unpartitioned())
+                  .withPath(cfd.fileName)
+                  .withFormat(cfd.fileFormat)
+                  .withFileSizeInBytes(cfd.length)
+                  .withMetrics(cfd.serializableMetrics.metrics());
+              DataFile dataFile = builder.build();
+              return dataFile;
+            }));
+          }
+        }
+
+        // Appending data files to the table
+        AppendFiles append = table.newAppend();
```

Comment:
How do we know that the write is always an append? Does Hive pass the operation so we can determine whether to use an overwrite or replacePartitions operation?

Reply:
Integrating with Hive update/delete statements would be a harder nut to crack.
Currently we are aiming only for inserts.
Does this still mean we have to think about replacePartitions, or overwrite? Or my assumption was correct that we need only to use newAppend in this case?

Reply:
The current patch is only aiming to insert new data to the table (no delete/update etc at the moment).
@rdblue: Is it enough to add data with newAppend, or we have to think about overwrite/replacePartitions too?
Thanks,
Peter